adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,888 @@
+                    ret = bind(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+                    
+                    // LogInfo("bind return: %d", ret);
+                    if (ret != 0) {
+                        result = __LINE__;
+                        LogError("bind socket failed");
+                    }
+                    else
+                    {
+                        LogInfo("bind socket OK");
+
+                        memset(&sock_addr, 0, sizeof(sock_addr));
+                        sock_addr.sin_family = AF_INET;
+                        sock_addr.sin_addr.s_addr = tls_io_instance->target_ip.addr;
+                        sock_addr.sin_port = htons(tls_io_instance->port);
+
+                        ret = connect(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+                        //LogInfo("connect return: %d %s", ret, ip_ntoa(&tls_io_instance->target_ip));
+                        //LogInfo("EINPROGRESS: %d", EINPROGRESS);
+                        if (ret == -1) {
+                            ret = lwip_net_errno(sock);
+                            LogInfo("lwip_net_errno ret: %d", ret);
+                            /* Codes_SRS_TLSIO_SSL_ESP8266_99_083: [ If connect and getsockopt failed, the tlsio_openssl_open shall return __LINE__. ] */
+                            if (ret != 115) { // EINPROGRESS
+                                result = __LINE__;
+                                ret = -1;
+                                LogError("socket connect failed, not EINPROGRESS %s", tls_io_instance->hostname);
+                            }
+                        }
+
+                        if(ret != -1)
+                        {
+                            FD_ZERO(&readset);
+                            FD_ZERO(&writeset);
+                            FD_ZERO(&errset);
+
+                            FD_SET(sock, &readset);
+                            FD_SET(sock, &writeset);
+                            FD_SET(sock, &errset);
+
+                            ret = lwip_select(sock + 1, NULL, &writeset, &errset, NULL);
+                            if (ret <= 0) {
+                                result = __LINE__;
+                                LogError("select failed: %d", lwip_net_errno(sock));
+                            } else 
+                            {
+                                if (!FD_ISSET(sock, &writeset) || FD_ISSET(sock, &errset)) {
+                                    result = __LINE__;
+                                    LogError("socket Error: %d", lwip_net_errno(sock));
+                                }else
+                                {
+                                    {
+                                        LogInfo("Socket Connect OK");
+                                        /* Codes_SRS_TLSIO_SSL_ESP8266_99_087: [ If SSL_new failed, the tlsio_openssl_open shall return __LINE__. ] */
+                                        ssl = SSL_new(ctx);
```

Comment:
Thanks for the feedback @JetstreamRoySprowl 

In `destroy_openssl_instance`, I always free this if it is not NULL.

```
if (tls_io_instance->ssl != NULL)
        {
            SSL_free(tls_io_instance->ssl);
            tls_io_instance->ssl = NULL;
            LogInfo("SSL_free");
        }
```